tests/test_models.py

```diff
@@ -20,28 +21,57 @@ def test_trick():
     assert t is Player.WEST
     assert trick.winner == Player.WEST
 
+
 def test_Hand():
-    hand = Hand(cards=[Card(CardSuit.SPADE, CardRank.ACE), 
-                       Card(CardSuit.HEART, CardRank.ACE), 
-                       Card(CardSuit.DIAMOND, CardRank.ACE),
-                       Card(CardSuit.SPADE, CardRank.KING)]);
-    
-    candidates = hand.candidate_cards();
-    assert len(candidates) == 4;
-    
-    candidates = hand.candidate_cards(CardSuit.SPADE);
-    assert len(candidates) == 2;
-    
-    candidates = hand.candidate_cards(CardSuit.CLUB);
-    assert len(candidates) == 4;
-    
-    hand.play_card(Card(CardSuit.HEART, CardRank.ACE));
-    candidates = hand.candidate_cards(CardSuit.HEART);
-    assert Card(CardSuit.HEART, CardRank.ACE) not in candidates;
-    assert len(candidates) == 3;
-    
-    hand.reverse();
-    candidates = hand.candidate_cards(CardSuit.HEART);
-    assert len(candidates) == 1;
-    assert Card(CardSuit.HEART, CardRank.ACE) in candidates;
-    
\ No newline at end of file
+    SA = Card(CardSuit.SPADE, CardRank.ACE)
+    SK = Card(CardSuit.SPADE, CardRank.KING)
+    HA = Card(CardSuit.HEART, CardRank.ACE)
+    DA = Card(CardSuit.DIAMOND, CardRank.ACE)
+    hand = Hand(cards=[SA, SK, HA, DA])
+
+    # function candidate_cards
+    candidates = hand.candidate_cards()
+    assert len(candidates) == 4
+
+    candidates = hand.candidate_cards(CardSuit.SPADE)
+    assert len(candidates) == 2
+
+    candidates = hand.candidate_cards(CardSuit.HEART)
+    assert len(candidates) == 1
+
+    candidates = hand.candidate_cards(CardSuit.DIAMOND)
+    assert len(candidates) == 1
+
+    candidates = hand.candidate_cards(CardSuit.CLUB)
+    assert len(candidates) == 4
+
+    # function play_card
+    hand.play_card(HA)
+    candidates = hand.candidate_cards(CardSuit.HEART)
+    assert HA not in candidates
+    assert len(candidates) == 3
+
+    # function reverse
+    hand.reverse()
+    candidates = hand.candidate_cards(CardSuit.HEART)
+    assert len(candidates) == 1
+    assert HA in candidates
+
+
+def test_Hand_empty():
+    hand = Hand()
+    candidates = hand.candidate_cards()
+    assert len(candidates) == 0
+
+    with pytest.raises(ValueError):
+        hand.play_card(Card(CardSuit.SPADE, CardRank.ACE))
+
+    with pytest.raises(IndexError):
+        hand.reverse()
+
+
+@pytest.mark.xfail
```

Comment:
We haven't had a duplicate prevention for method add_card in Hand class, right?

Reply:
Yep, so I add the testcase and mark it as xfail. This mean we expect the testcase will fail.
After we implement the function, we will remove the mark.

Reply:
Got it